### ISSUE TITLE
Storage exception on file upload to an NFS mount

### ISSUE THREAD

**Issue by MarinBernard** [omeka/Omeka#934]
Hi,

I'm trying to tweak my Omeka install (EL8 + Apache 2.4 + PHP-FPM 7.4) to make it store original file uploads on an NFS mount. The `files/original` directory was symlinked to a directory within an NFS mount. The apache process is able to read and write data on the NFS path, and Omeka successfully serves original uploads from there. However, any file upload will fail with the following exception:

````
Omeka_Storage_Exception: Unable to store file. in /var/www/html/application/libraries/Omeka/Storage/Adapter/Filesystem.php:119
Stack trace:
#0 /var/www/html/application/libraries/Omeka/Storage.php(67): Omeka_Storage_Adapter_Filesystem->store()
#1 /var/www/html/application/models/File.php(424): Omeka_Storage->__call()
#2 /var/www/html/application/models/Job/FileProcessUpload.php(23): File->storeFiles()
#3 /var/www/html/application/libraries/Omeka/Job/Dispatcher/Adapter/Synchronous.php(25): Job_FileProcessUpload->perform()
#4 /var/www/html/application/libraries/Omeka/Job/Dispatcher/Default.php(137): Omeka_Job_Dispatcher_Adapter_Synchronous->send()
#5 /var/www/html/application/models/File.php(203): Omeka_Job_Dispatcher_Default->send()
#6 /var/www/html/application/libraries/Omeka/Record/AbstractRecord.php(280): File->afterSave()
#7 /var/www/html/application/libraries/Omeka/Record/AbstractRecord.php(548): Omeka_Record_AbstractRecord->runCallbacks()
#8 /var/www/html/application/models/Item.php(336): Omeka_Record_AbstractRecord->save()
#9 /var/www/html/application/models/Item.php(249): Item->saveFiles()
#10 /var/www/html/application/libraries/Omeka/Record/AbstractRecord.php(280): Item->afterSave()
#11 /var/www/html/application/libraries/Omeka/Record/AbstractRecord.php(548): Omeka_Record_AbstractRecord->runCallbacks()
#12 /var/www/html/application/libraries/Omeka/Controller/AbstractActionController.php(188): Omeka_Record_AbstractRecord->save()
#13 /var/www/html/application/controllers/ItemsController.php(148): Omeka_Controller_AbstractActionController->addAction()
#14 /var/www/html/application/libraries/Zend/Controller/Action.php(516): ItemsController->addAction()
#15 /var/www/html/application/libraries/Zend/Controller/Dispatcher/Standard.php(308): Zend_Controller_Action->dispatch()
#16 /var/www/html/application/libraries/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch()
#17 /var/www/html/application/libraries/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#18 /var/www/html/application/libraries/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#19 /var/www/html/application/libraries/Omeka/Application.php(73): Zend_Application->run()
#20 /var/www/html/admin/index.php(28): Omeka_Application->run()
#21 {main}
````

Which is linked to the following entry in the PHP-FPM error log:

````
[23-Jan-2021 20:57:35 UTC] PHP Warning:  rename(/tmp/ceb986d0f1c615aa630bd9cac34c903d.png,/var/www/html/files/original/ceb986d0f1c615aa630bd9cac34c903d.png): Invalid argument in /var/www/html/application/libraries/Omeka/Storage/Adapter/Filesystem.php on line 222
````

I was able to fix file uploads by replacing the final statement in the [_rename()](https://github.com/omeka/Omeka/blob/261631eaceac81670c2b008051cfa11042cebb9c/application/libraries/Omeka/Storage/Adapter/Filesystem.php#L215) method of the `Omeka_Storage_Adapter_Filesystem ` class:

````php
return rename($source, $dest);
````

with

````php
return copy($source, $dest) && unlink($source);
````

Uploading to a symlinked directory works fine, at least when the target directory is located on the same file system as the symlink. `rename()` is [known to have issues](https://www.php.net/manual/en/function.rename.php#113943) renaming directories crossing file system boundaries, but this should'nt be a problem there because we're only dealing with files here.

`rename()` also has [problems](https://www.php.net/manual/en/function.rename.php#117590) with file systems which do not allow `chmod()` or  `chown()` operations. This is the case here: we use NFS4 with ACLs, and apache is not allowed to change file mode or owner. In such contexts, a recommended workaround is to use `copy()` and  `unlink()` instead of `rename()`.

The _invalid argument_ error remains a mystery, though. The function only accepts a source and a destination, and both logged strings seem to be valid paths.

While I was debugging the error, I tried to call `copy()` only, without `unlink()`. I hoped this would reveal the presence of a temporary file in `/tmp`, but it did not. I suppose some other piece of code takes care of removing uploaded files there. If so, would it be possible to simply replace `rename()` with `copy()` and get rid of `rename()`'s limitations regarding NFS ?

Thanks for your insights!

**Comment by zerocrates** [comment]
Interesting that we haven't really seen this issue before, though we have relatively few users using separate filesystems for file storage and even fewer using NFS, so I guess it's not surprising really.

The "invalid argument" error here is just coming from PHP's attempt to "simulate" a rename using copy/chmod/chown/unlink. Some people reporting this same issue say that the copy actually succeeds as its done first, but rename returns false when one of the other operations fails and the warning is printed. The warning isn't related to the actual arguments we pass to rename().

I think the most likely way forward here would just be adding a configuration option to the filesystem storage adapter to switch to copy-then-unlink. Rename is much more efficient as an operation for users who _are_ on the same filesystem, and that kind of difference could add up quickly in scenarios like imports.

As for needing unlink, are you sure about how you checked for the remaining temporary file? There is no process that would delete the original temporary file, unless there's an error during the ingest process.

**Comment by MarinBernard** [comment]
Thanks for your answer. Indeed, in our setup, any call to `chown()` or `chmod()` is doomed to fail because:

1. We use NFSv4 ACLs which do not allow `chmod` operations
2. We use kerberized NFS which, implies to map the identity under which the Apache process is running (the `apache` local user) to a centrally-managed Kerberos principal

As a consequence, while the web server is allowed to perform regular read/write operations on the share, it won't be able to change file modes, nor to take ownership of the files. Files will always appear as being owned by the Kerberos identity, not the local one. Using `copy()` + `unlink()` fixes the issue since we don't even try to enforce file attributes.

Yes, adding a config option to switch upload methods would be great!

Regarding `unlink()` I tried again and it's my mistake: I assumed `/tmp` was my temporary upload directory whereas systemd private tmp dirs are in use. I made a test again and indeed, without `unlink()`, uploaded files remain present in the tmp dir.
